### PR TITLE
Release candidate #1 pre-release PR for Fall24 meta-release

### PR DESCRIPTION
**Pull request by jpengar**
#### What type of PR is this?

* subproject management

#### What this PR does / why we need it:

First release candidate for v0.4.0 pre-release PR for Fall24 meta-release

#### Which issue(s) this PR fixes:

Related to https://github.com/camaraproject/HomeDevicesQoD/issues/63

#### Special notes for reviewers:

N/A

#### Changelog input

N/A

#### Additional documentation 

N/A

### CODE REVIEW

**Comment by tanjadegroot** [comment]
Review of home_devices_qod.yaml:

In the Info Object: remove the fields:
    termsOfService: http://swagger.io/terms/
    contact:

the externalDocs url in most API files points to the actual repository, so you could do that as well.

**Comment by tanjadegroot** [comment]
review of README.md:

some rewordings:
-  first bullet: Pre-release 0.4.0-rc.1 is available here --> Pre-release r1.1 with API version 0.4.0-rc.1 is available here.
- second bullet: I would rephrase "latest public release" to "the previous release"

**Review by tanjadegroot — APPROVED** [review]

**Comment by tanjadegroot** [comment]
Review of CHANGELOG.md: all good

File: README.md [hunk 1]
```diff
@@ -18,6 +18,12 @@ Repository to describe, develop, document and test the HomeDevicesQoD API family
 
 ## Release Information
 * Note: Please be aware that the project will have updates to the main branch. There are no compatibility guarantees associated with code in any branch, including main, until a new release is created. For example, changes may be reverted before a release is created. **For best results, use the latest available release**.
+* `NEW`: Pre-release `0.4.0-rc.1` is available [here](https://github.com/camaraproject/HomeDevicesQoD/tree/r1.1)
```
[the remark this answers is not in the document]
Comment (jpengar) [reply README.md]: Fixed in https://github.com/camaraproject/HomeDevicesQoD/pull/68/commits/a53d0255f547685a01552df6a31d9c38678bf7bb

**Comment by jpengar** [comment]
> Review of home_devices_qod.yaml:
> 
> In the Info Object: remove the fields: termsOfService: http://swagger.io/terms/ contact:
> 
> the externalDocs url in most API files points to the actual repository, so you could do that as well.
>

Fixed in https://github.com/camaraproject/HomeDevicesQoD/pull/68/commits/a53d0255f547685a01552df6a31d9c38678bf7bb
 
> the documentation uses "Telco speak":
> 
> * Service Enabling Network Function API may not be understood by developers
> * DSCP, Service Class etc. is quite technical/network focused.
>   The section on API functionality would be a good starting section.

In fact, the API was originally designed to only use DSCP as a parameter of the request, and the API spec was adapted to be more developer friendly by incorporating the service class concept of RFC4594. More context in:  https://github.com/camaraproject/HomeDevicesQoD/issues/23 && https://github.com/camaraproject/HomeDevicesQoD/pull/25

> The documentation does not cover explanation about the error cases, or the concepts of error cases: maximum nr of devices, home network occupancy, etc. this is important for the developer

Errors are documented as Examples following the  Commonalities definitions. This was previously addapted for this meta-release in https://github.com/camaraproject/HomeDevicesQoD/pull/64

**Comment by tanjadegroot** [comment]
Review of test file:
scenario line 33 
- it seems that it lists "serviceClass" as one of the possible example values. Is that the intention or is it meant to be a header of the list ?
- does not list "standard" as one of the possible example options.

**Comment by jpengar** [comment]
> Review of test file: scenario line 33
> 
> * it seems that it lists "serviceClass" as one of the possible example values. Is that the intention or is it meant to be a header of the list ?

"serviceClass" is the header of the Examples table. It follows the Gherkin language syntax for "Scenario Outline": https://cucumber.io/docs/gherkin/reference/#scenario-outline

> * does not list "standard" as one of the possible example options.

The test `@setQos_01_set_service_class_success` is intended to test all possible values of "serviceClass" for the IP traffic corresponding to a specific user home device (except "standard", which is the default value). Then the next test case `@setQos_02_restore_default_service_class_success` tests to restore the `standard` value from all possible values of "serviceClass". Exactly the other way around.

**Review by fernandopradocabrillo — APPROVED** [review]

**Comment by tanjadegroot** [comment]
Review of API readiness checklist: all OK

On the error documentation, to my understanding it is mandated by the https://github.com/camaraproject/Commonalities/blob/main/documentation/API-DocumentationTemplate.md.
We can leave it for the M4 - if you could create an issue to not forget that would be perfect..

**Comment by tanjadegroot** [comment]
LGTM from ReleaseManagement view.

**Review by tanjadegroot — APPROVED** [review]
For M4, need to add description of errors in the API specification.

LGTM from Release Management

**Comment by jpengar** [comment]
> Review of API readiness checklist: all OK
> 
> On the error documentation, to my understanding it is mandated by the https://github.com/camaraproject/Commonalities/blob/main/documentation/API-DocumentationTemplate.md. We can leave it for the M4 - if you could create an issue to not forget that would be perfect..

@tanjadegroot Could you point me to some other APIs that do what you expect here in terms of errors? As I understand it, errors are documented as part of each endpoint definition and they are documented as examples with descriptions:

e.g.
```
examples:
            Conflict:
              value:
                status: 409
                code: CONFLICT
                message: Conflict with the current state of the target resource
            TooManyDevices:
              value:
                status: 409
                code: HOME_DEVICES_QOD.TOO_MANY_DEVICES
                message: Exceeded the maximum quantity of devices with non-default QoS behaviour
            RssiBelowThreshold:
              value:
                status: 409
                code: HOME_DEVICES_QOD.RSSI_BELOW_THRESHOLD
                message: RSSI from device is below allowed threshold
            QosTooHigh:
              value:
                status: 409
                code: HOME_DEVICES_QOD.QOS_TOO_HIGH
                message: Requested QoS is higher than the maximum QoS allowed
            OccupancyAboveThreshold:
              value:
                status: 409
                code: HOME_DEVICES_QOD.OCCUPANCY_ABOVE_THRESHOLD
                message: The occupancy is above the allowed threshold
            NotConnectedToRequiredInterface:
              value:
                status: 409
                code: HOME_DEVICES_QOD.NOT_CONNECTED_TO_REQUIRED_INTERFACE
                message: Device is not connected to the required interface (e.g. WiFi 5GHz interface)
            NotSupportedRequiredInterface:
              value:
                status: 409
                code: HOME_DEVICES_QOD.NOT_SUPPORTED_REQUIRED_INTERFACE
                message: Device does not support required interface (e.g. WiFi 5GHz interface)
            QosAlreadySetToDefault:
              value:
                status: 409
                code: HOME_DEVICES_QOD.QOS_ALREADY_SET_TO_DEFAULT
                message: Device QoS is already set to default value
```
Other APIs like QualityOnDemand do the same thing... e.g. https://redocly.github.io/redoc/?url=https://raw.githubusercontent.com/camaraproject/QualityOnDemand/release-0.10.1/code/API_definitions/qod-api.yaml&nocors

The same way that the following part of the documentation is documented and detailed on each endpoint definition, not in info.description:
image

e.g:
image

**Comment by jpengar** [comment]
@tanjadegroot Please take a look at https://github.com/camaraproject/HomeDevicesQoD/pull/68#issuecomment-2257645191. I think errors are documented as expected and there would be no missing inline doc. https://github.com/camaraproject/Commonalities/blob/main/documentation/API-DocumentationTemplate.md is a bit outdated IMHO. Thanks a lot in advance!

Just to double-check before I merge the PR :)

**Comment by jpengar** [comment]
@tanjadegroot @fernandopradocabrillo https://github.com/camaraproject/HomeDevicesQoD/pull/68/commits/a828df35df443d5d64ca35f4cc0e5573a30e22c4 is needed to address last changes in info.description template in ICM r0.2.0-rc.2 to fix link references. Further info in: https://github.com/camaraproject/IdentityAndConsentManagement/issues/185

**Review by hdamker — APPROVED** [review]

**Review by fernandopradocabrillo — APPROVED** [review]

**Comment by hdamker** [comment]
@tanjadegroot I just approved the change done in https://github.com/camaraproject/HomeDevicesQoD/commit/a828df35df443d5d64ca35f4cc0e5573a30e22c4. The final saying from Release Management is still on you.

**Comment by jpengar** [comment]
@tanjadegroot May I merge the PR then?

**Comment by hdamker** [comment]
> @tanjadegroot May I merge the PR then?

As @tanjadegroot is now on vacation I took over the release review issue. Looks good from release management perspective.

Minor comment, which you can ignore: #62 and #67 are not really needed within the CHANGELOG.md, as they are about subproject management and not relevant for API users.

**Review by hdamker — APPROVED** [review]
LGTM from release management.
